// @framer/codemilli.twitter-sticky/code/App.tsx
import { Data, animate, Override, Animatable } from "framer";

export let store = {};
export const starFrameData = Data({
  scale: 1
});

function setState(id) {
  if (!store[id]) {
    store[id] = {};
    store[id].skeletonData = Data({
      scale: Animatable(1),
      opacity: Animatable(1)
    });
    store[id].starData = Data({
      scale: Animatable(1),
      opacity: Animatable(0)
    });
  }
  return store[id];
}

export const Touchable: Override = props => {
  const { skeletonData, starData } = setState(1);
  return {
    onTap() {
      skeletonData.opacity.set(0);
      animate.spring(starData.opacity, 1);
      animate.spring(starData.scale, 10);
    }
  };
};

export const SkeletonStar: Override = props => {
  const { skeletonData } = setState(1);
  return {
    scale: skeletonData.scale,
    opacity: skeletonData.opacity
  };
};

export const Star: Override = props => {
  const { starData } = setState(1);
  return {
    scale: starData.scale,
    opacity: starData.opacity
  };
};

export const StarFrame: Override = props => {
  return {
    scale: starFrameData.scale
  };
};

export const ScaleSlider: Override = props => {
  return {
    onValueChange: (value: number) => {
      starFrameData.scale = value / 10;
    }
  };
};
